ocaml/sdk-gen/common/CommonFunctions.ml

```diff
@@ -111,137 +93,188 @@ and is_method_static message =
+      let index2 = list_index_of y code_name_order in
+      index1 - index2
+  | None, _ ->
+      -1
+  | _, None ->
+      1
+  | Some v1, Some v2 ->
+      if v1.major == v2.major && v1.minor == v2.minor && v1.micro == v2.micro
+      then
+        0
+      else if v1.major == v2.major && v1.minor == v2.minor then
+        v1.micro - v2.micro
+      else if v1.major == v2.major then
+        v1.minor - v2.minor
+      else
+        v1.major - v2.major
+
+let rec lifecycle_matcher milestone lifecycle =
+  match lifecycle with
+  | [] ->
+      ""
+  | (x, y, _) :: tl ->
+      if x == milestone then
+        y
+      else
+        lifecycle_matcher milestone tl
+
+let get_published_release_for_param releases =
+  let filtered =
     List.filter
-      (fun x -> match x with [] -> false | _ -> true)
-      (getValid paramGroups)
+      (fun x -> x <> "closed" && x <> "3.0.3" && x <> "debug")
+      releases
   in
-  if not (expRelease = "") then
-    [params]
-  else
-    list_distinct filteredGroups
+  match filtered with [] -> "" | hd :: _ -> hd
+
+let get_prototyped_release lifecycle = lifecycle_matcher Prototyped lifecycle
 
-and get_release_name codename =
+let get_published_release lifecycle = lifecycle_matcher Published lifecycle
+
+let get_release_branding codename =
   try
     let found =
       List.find (fun x -> code_name_of_release x = codename) release_order
     in
     found.branding
-  with Not_found -> ""
-
-and get_first_release releases =
-  let filtered = List.filter (fun x -> List.mem x releases) code_name_order in
-  match filtered with [] -> "" | hd :: _ -> hd
+  with Not_found -> codename
 
-and get_first_release_string release =
-  if release = "" then
-    ""
-  else
-    sprintf "First published in %s." (get_release_name release)
+let group_params_per_release params =
+  let same_release p1 p2 =
+    compare_versions
+      (get_published_release_for_param p1.param_release.internal)
+      (get_published_release_for_param p2.param_release.internal)
+    = 0
+  in
+  let rec groupByRelease acc = function
+    | [] ->
+        acc
+    | hd :: tl ->
+        let l1, l2 = List.partition (same_release hd) tl in
+        groupByRelease ((hd :: l1) :: acc) l2
+  in
+  List.rev (groupByRelease [] params)
 
-and get_deprecated_info_message_string version =
-  match version with
-  | None ->
-      ""
-  | Some x ->
-      sprintf "Deprecated since %s." (get_release_name x)
+let collate l =
+  let rec collator x acc = function
+    | [] ->
+        acc
+    | hd :: tl ->
+        collator (hd :: x) ((hd :: x) :: acc) tl
+  in
+  collator [] [] l
 
-and get_prototyped_release lifecycle =
-  match lifecycle with [(Prototyped, release, _)] -> release | _ -> ""
+let gen_param_groups message params =
+  let expRelease = get_prototyped_release message.msg_lifecycle in
+  let paramGroups = group_params_per_release params in
+  let overloadGroups =
+    List.rev (List.map List.concat (List.map List.rev (collate paramGroups)))
```

Comment:
This is quite a mouthful, and it's tough to see what is `collate` doing, I'd like to try to come up with something easier to understand. Could you try to explain the relationship between parameter groups and overload groups?

Reply:
Although modern versions of C# support optional parameters, I wanted to avoid making too many changes to the existing C# generator, and also wanted to be able to reuse at least some of the code with the generator for java, which does not support optional parameters in a straight forward way, so I kept the generation of traditional method overloads.

This means that if the Ocaml call has a number of parameters added at 4 different releases, e.g.
`call(param_rio, param_rio, param_lima, param_1_302_0, param_1_302_0, param_22_1_18)`
group_params_per_release splits the above parameters in 4 groups and then the "mouthful" line combines them so this results to the following c#/java overloads:
`call(param_rio, param_rio)`
`call(param_rio, param_rio, param_lima)`
`call(param_rio, param_rio, param_lima, param_1_302_0, param_1_302_0)`
`call(param_rio, param_rio, param_lima, param_1_302_0, param_1_302_0, param_22_1_18)`